File: Algorithms/Python3.x/541-Reverse_String_II.py
# Runtime: 32 ms, faster than 95.75% of Python3 online submissions for Reverse String II.
# Memory Usage: 12.9 MB, less than 100.00% of Python3 online submissions for Reverse String II.
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        s = list(s)
        for start in range(0, len(s), 2 * k):
            end = min(start + k - 1, len(s) - 1)
            while start < end:
                s[start], s[end] = s[end], s[start]
                start += 1
                end -= 1
        return ''.join(s)


# direct
# Runtime: 32 ms, faster than 95.75% of Python3 online submissions for Reverse String II.
# Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Reverse String II.
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        s = list(s)
        for i in range(0, len(s), 2 * k):
            s[i:i + k] = reversed(s[i:i + k])
        return ''.join(s)

File: Algorithms/C++/541-Reverse_String_II.cpp
// Runtime: 4 ms, faster than 97.48% of C++ online submissions for Reverse String II.
// Memory Usage: 9.6 MB, less than 60.00% of C++ online submissions for Reverse String II.
class Solution {
public:
    string reverseStr(string s, int k) {
        for (int start = 0; start < s.size(); start += 2 * k) {
            int left = start, right = min(left + k - 1, static_cast<int>(s.size()) - 1);
            while (left < right) {
                swap(s[left], s[right]);
                left++;
                right--;
            }
        }
        return s;
    }
};
